android/src/main/java/com/reactnativespokestack/Downloader.kt

```diff
@@ -0,0 +1,30 @@
+package com.reactnativespokestack
+
+import android.content.Context
+import android.util.Log
+import java.io.File
+import java.io.FileOutputStream
+import java.net.URL
+
+class Downloader(private val context: Context) {
+  private val logTag = "Spokestack"
+
+  fun downloadModel(filename: String, url: String): String {
+    val file = File(context.cacheDir, filename)
+    if (file.exists() && file.canRead() && file.length() > 0) {
+      Log.d(logTag, "Returning existing file at ${file.canonicalPath}")
+      return file.canonicalPath
+    }
+    val cn = URL(url).openConnection()
```

Comment:
Interesting—RN must run non-UI code on a background thread by default, otherwise you'd get smacked down by the runtime doing it this way. This should work and is pretty similar to what the android tray does, but you should call `out.close()` when you're done downloading.

If you wanted it to be parallel like Swift, you could always dig into doing [something with coroutines](https://jivimberg.io/blog/2018/05/04/parallel-map-in-kotlin/).

Reply:
I didn't realize it was like async/await. That could be worth it.